server/src/main/java/org/elasticsearch/cluster/ClusterState.java

```diff
@@ -49,25 +56,45 @@
- * by the {@link ClusterService}. After every update the {@link ClusterStatePublisher#publish} method publishes a new version of the cluster
- * state to all other nodes in the cluster.
+ * Conceptually immutable, but in practice it has a few components like {@link RoutingNodes} which are pure functions of the immutable state
+ * but are expensive to compute so they are built on-demand if needed.
  * <p>
- * Implements the {@link Diffable} interface in order to support publishing of cluster state differences instead of the entire state on each
- * change. The publishing mechanism only sends differences to a node if this node was present in the previous version of the cluster state.
- * If a node was not present in the previous version of the cluster state, this node is unlikely to have the previous cluster state version
- * and should be sent a complete version. In order to make sure that the differences are applied to the correct version of the cluster
- * state, each cluster state version update generates {@link #stateUUID} that uniquely identifies this version of the state. This uuid is
- * verified by the {@link ClusterStateDiff#apply} method to make sure that the correct diffs are applied. If uuids don’t match, the {@link
- * ClusterStateDiff#apply} method throws the {@link IncompatibleClusterStateVersionException}, which causes the publishing mechanism to send
- * a full version of the cluster state to the node on which this exception was thrown.
+ * The {@link Metadata} portion is written to disk on each update so it persists across full-cluster restarts. The rest of this data is
+ * maintained only in-memory and resets back to its initial state on a full-cluster restart, but it is held on all nodes so it persists
+ * across master elections (and therefore is preserved in a rolling restart).
+ * <p>
+ * Updates are triggered by submitting tasks to the {@link MasterService} on the elected master, typically using a {@link
+ * TransportMasterNodeAction} to route a request to the master on which the task is submitted with {@link
+ * ClusterService#submitStateUpdateTask}. Submitted tasks have an associated {@link ClusterStateTaskConfig} which defines its priority and a
+ * timeout. Tasks are processed in priority order, so a flood of higher-priority tasks can starve lower-priority ones of access to the
+ * master. Use priorities other than {@link Priority#NORMAL} with care. Tasks associated with client actions should typically have a
+ * timeout, or otherwise be sensitive to client cancellations, to avoid surprises caused by the execution of stale tasks long after they are
+ * submitted (since clients tend to time out). In contrast internal tasks should generally prefer an infinite timeout, especially if a
+ * timeout would simply trigger a retry.
+ * <p>
+ * Each batch of tasks yields a new {@link ClusterState} which is published to the cluster by {@link ClusterStatePublisher#publish}.
```

Comment:
I am not sure if I follow "Each batch of tasks" , a batch of any tasks ? and why a batch ?

Maybe this would be a good place to recommend processing them in batch ?

Reply:
I added a sentence to explain where batches come from but struggled to find a way to recommend batching without breaking up the flow too much. However there are some API changes incoming (#82383 is a start) that will make it harder to avoid batching so I think it's ok to leave that out.

Reply:
Thanks, the latest update reads much better.